### DISCUSSION TITLE
ESW7010 Warming Drawer

### DISCUSSION THREAD

**Discussion by Ursli71** [astrandb/miele#433] [Q&A]
Hi,
I like to integrate my Miele Warming Drawer into my HA automation. The Miele Ingration nicely found the drawer and a basic integration took place. It only created 3 sensors as per the attached screenshot:     
![Screenshot 2024-02-19 161123](https://github.com/astrandb/miele/assets/156338250/681d740b-71ff-4a6a-9332-5aa36e567577)

What it missed are more sensors for the drawer in HA similar to what the Miele API ('https://api.mcs3.miele.com/v1/devices/{deviceID}/state')  is offering, eg ProgramID, ProgramPhase, remaining time, etc. See the dump from the Miele API:
{
  "ProgramID": {
    "value_raw": 2,
    "value_localized": "Warm Dishes/Plates",
    "key_localized": "Program name"
  },
  "status": {
    "value_raw": 5,
    "value_localized": "In use",
    "key_localized": "status"
  },
  "programType": {
    "value_raw": 0,
    "value_localized": "Program",
    "key_localized": "Program type"
  },
  "programPhase": {
    "value_raw": 3073,
    "value_localized": "Heating-up phase",
    "key_localized": "Program phase"
  },
  "remainingTime": [
    0,
    0
  ],
  "startTime": [
    0,
    0
  ],
  "targetTemperature": [
    {
      "value_raw": 7000,
      "value_localized": 70,
      "unit": "Celsius"
    }
  ],
  "coreTargetTemperature": [],
  "temperature": [
    {
      "value_raw": 5158,
      "value_localized": 51.58,
      "unit": "Celsius"
    },
    {
      "value_raw": -32768,
      "value_localized": null,
      "unit": "Celsius"
    },
    {
      "value_raw": -32768,
      "value_localized": null,
      "unit": "Celsius"
    }
  ],
  "coreTemperature": [],
  "signalInfo": false,
  "signalFailure": false,
  "signalDoor": false,
  "remoteEnable": {
    "fullRemoteControl": true,
    "smartGrid": false,
    "mobileStart": true
  },
  "ambientLight": null,
  "light": null,
  "elapsedTime": [],
  "spinningSpeed": {
    "unit": "rpm",
    "value_raw": null,
    "value_localized": null,
    "key_localized": "Spin speed"
  },
  "dryingStep": {
    "value_raw": null,
    "value_localized": "",
    "key_localized": "Drying level"
  },
  "ventilationStep": {
    "value_raw": null,
    "value_localized": "",
    "key_localized": "Fan level"
  },
  "plateStep": [],
  "ecoFeedback": null,
  "batteryLevel": null
}

Now the question is how can I get access to those more states of the drawer so that I can use them for my UI and automation? 

Any input is welcome.
Cheers, Urs

**Comment by astrandb** [comment]
Thank you for your report. 
The warming drawer is not supported in current version of the integration as you seem to be the first integration user with such a device. I can enable some basic support based on the data you sent. However, it might be that there are programs and program phases that are specific to this device type and then will be displayed with a number instead of text.
I'll be back...

**Comment by astrandb** [comment]
I was quick today!
There is now a beta version 2024.2.0b0 available in HACS. Select redownload of the Miele integration and toggle Show beta versions and proceed.
All feed-back is appreciated.

**Reply by Ursli71** [reply]
Dear Åke,
Wow indeed - super quick! Let me download the beta and I report back. 
THANKS & cheers
Urs

**Reply by Ursli71** [reply]
I tried to be quick as well and did so immediately. I can confirm that I have now all sensors integrated and they are working overall fine. Following detailed observations:
- Indeed the program is shown as numbers. This is fine for me as I can handle it via the UI myself. If interested I am happy to share the  4 programs the drawer has, if you like to include them in one of the next versions
- the started_at, start_at and start_time seem not to work, but not sure if the API is correct and the drawer has this functionality. In the Miele app itself, I can NOT choose a start time - I found out just now. :-)
-  elapsed_time is the same. The Miele app shows me not the elapsed_time
- But the App shows me a time left. So in the API this would probably be the "remainingTime". So might be something for future versions as well. 
See screenshot of all the new sensors:
![Screenshot 2024-02-19 193944](https://github.com/astrandb/miele/assets/156338250/951656df-1074-4f5e-9009-1ddf83525010)

And again thanks for super quick turnaround and support! 
Best regards to Sweden from Switzerland

**Reply by astrandb** [reply]
The documentation from Miele is sparse and not always correct so some reverse engineering is needed.

One question: Is the door sensor working?

If you can find out english program names from sensor Program and program phases it would be super. Translation to other languages are handled in a separate process.

I was a bit generous when enabling sensors in this beta and I think we should remove non-working sensors before formal release.

**Reply by Ursli71** [reply]
Now as I progress with the drawer integration (or rather all my Miele appliances - I will ask later another question on oven and steamer integration) I realize that executing actions as well programs is probably another piece needed to have the drawer fully enabled in your Miele HA integration. Especially the set target temperature is the one I am after. 

I tried to do so via raw action, but got a 400:
serialno: 123456
extra:
  targetTemperature:
    value: 70

 I assume it would be possible, but I guess there are many more amateurs like me and thus here the API dumps.
Get actions:
{
  "processAction": [],
  "light": [],
  "ambientLight": [],
  "startTime": [],
  "ventilationStep": [],
  "programId": [
    1,
    2,
    3,
    4
  ],
  "targetTemperature": [],
  "deviceName": true,
  "powerOn": true,
  "powerOff": false,
  "colors": [],
  "modes": [],
  "runOnTime": []
}

And the 4 programs for the drawer:
[
  {
    "programId": 1,
    "program": "Warm Cups/Glases",
    "parameters": {}
  },
  {
    "programId": 2,
    "program": "Warm Dishes/Plates",
    "parameters": {}
  },
  {
    "programId": 3,
    "program": "Keep Warm",
    "parameters": {}
  },
  {
    "programId": 4,
    "program": "Slow Roasting",
    "parameters": {}
  }
]
*recopied the 4 programs. As earlier only the 3 showed up (API only shows the not currently running ones)

**Comment by Ursli71** [comment]
> The documentation from Miele is sparse and not always correct so some reverse engineering is needed.
> 
> One question: Is the door sensor working?
> 
> If you can find out english program names from sensor Program and program phases it would be super. Translation to other languages are handled in a separate process.
> 
> I was a bit generous when enabling sensors in this beta and I think we should remove non-working sensors before formal release.

I can confirm the door sensor is working. Just tested it. And yes can imagine the hours of reverse engineering went into this piece of code. 

english names I have provided in the other requests on programs and actions. 

And logically agreed to remove not working sensors.

**Comment by astrandb** [comment]
```
"targetTemperature": [],
"deviceName": true,
"powerOn": true,
"powerOff": false,
"colors": [],
"modes": [],
"runOnTime": []
```
The actions api shows what can be controlled and not controlled in the state the appliance is at the moment.

- So you can now start program ID 1, 2, 3 and 4
- You can set powerOn (probably - not always true even if api says so)
- Target temperature cannot be set. Usually so for ovens but e.g. freezers and fridges accept target temperature.
- If you turn your device on, powerOff action will probably turn to true. 
- runOnTime is not supported.


Please turn the device on (manually) and compare the actions output with previous.

I can easily activate a power switch if you can confirm that the actions api works as expected.

**Reply by Ursli71** [reply]
ok, despite in the Miele App (see screenshot) I can control the temperature thus is not possible via API.  This is another example of the limitation of external vs internal API as I have read in another thread.  
![Screenshot 2024-02-20 at 10 04 02](https://github.com/astrandb/miele/assets/156338250/89e7261e-b038-4003-9a3f-3aab780b9f87)

Yes if I turn the device on (manually, via API or Miele App) the powerOff action turns to true. When turned on it automatically turns on the last program and starts heating. So yes a power switch can be implemented. I have done so already with yaml switches, but yeah via your integration is even better.

**Comment by astrandb** [comment]
> As earlier only the 3 showed up (API only shows the not currently running ones)

Yes, that is because the actions endpoint only show what state you can change to. Present states are none, false or absent.
There are also program phases, please try to find them also so can we have it all fixed for next beta.

**Reply by Ursli71** [reply]
so there are 4 program phases I found out:
3075: door open
3073: heating-up
3094: keeping warm
3088: cooling down

**Reply by astrandb** [reply]
OK, thanks. I will add those including the power switch later today - hopefully.

**Reply by Ursli71** [reply]
no worries - take your time and a big thank you anyway

**Comment by Ursli71** [comment]
Now I have a different question not related to the warming drawer, but I thought I post it here quickly as it's about Miele integration. I assume I found the answer already, but wanted to confirm with the expert. I searched the other threads, but did not find something which helped me. 

I have more Miele appliances, but with 2 I am not able to use the API properly:  Miele oven type H7860-60BP and steam combination oven type DGC7845. Simple switch on/off or light on/off as well sensors work nicely, but every time I want to set a program I get a "code": 400,
  "message": "Response from setting program was not OK or OK but RESPONSE_FROM_DEVICE_ERROR instead.". 

Now I did try it directly via Miele API (https://www.miele.com/developer/swagger-ui/index.html) and for both continuously I get the same error code. I reached out to Miele developer email, but for a month no response. So if you have any good idea on how I could get this to work, your input would be highly appreciated. 

If needed happy to create/move it to a separate discussion, but I assume that this is a Miele bug with those 2 applicances, but as said wanted to get your view on it and if as assumed then we can close this comment/question.  Thanks

**Comment by astrandb** [comment]
You could turn the drawer on and check the get programs endpoint on Swagger and see if you get any available programs.

**Reply by Ursli71** [reply]
your suggestion is related to my post on error code: 400? if so it's not related to the drawer, but to my oven and steamer. I cannot set programs for those, but fine for drawer, fridge and dishwasher. on Swagger I get the available programs, but no chance to set them via API...

**Comment by astrandb** [comment]
Please upload the output from Swagger with the available programs list

**Reply by Ursli71** [reply]
for the drawer? as per above:
[
{
"programId": 1,
"program": "Warm Cups/Glases",
"parameters": {}
},
{
"programId": 2,
"program": "Warm Dishes/Plates",
"parameters": {}
},
{
"programId": 3,
"program": "Keep Warm",
"parameters": {}
},
{
"programId": 4,
"program": "Slow Roasting",
"parameters": {}
}
]

**Comment by astrandb** [comment]
And oven, please.

**Reply by Ursli71** [reply]
ok for the steam combination oven type DGC7845 this is the dump:
[
  {
    "programId": 13,
    "program": "Fan plus",
    "parameters": {
      "temperature": {
        "min": 30,
        "max": 225,
        "step": 5,
        "mandatory": false
      },
      "duration": {
        "min": [
          0,
          1
        ],
        "max": [
          12,
          0
        ],
        "mandatory": false
      }
    }
  },
  {
    "programId": 24,
    "program": "Conventional heat",
    "parameters": {
      "temperature": {
        "min": 30,
        "max": 225,
        "step": 5,
        "mandatory": false
      },
      "duration": {
        "min": [
          0,
          1
        ],
        "max": [
          12,
          0
        ],
        "mandatory": false
      }
    }
  },
  {
    "programId": 8,
    "program": "Steam cooking",
    "parameters": {
      "temperature": {
        "min": 40,
        "max": 100,
        "step": 5,
        "mandatory": false
      },
      "duration": {
        "min": [
          0,
          1
        ],
        "max": [
          10,
          0
        ],
        "mandatory": false
      }
    }
  },
  {
    "programId": 72,
    "program": "Sous-vide",
    "parameters": {
      "temperature": {
        "min": 45,
        "max": 90,
        "step": 1,
        "mandatory": false
      },
      "duration": {
        "min": [
          0,
          1
        ],
        "max": [
          10,
          0
        ],
        "mandatory": false
      }
    }
  },
  {
    "programId": 14,
    "program": "Intensive bake",
    "parameters": {
      "temperature": {
        "min": 50,
        "max": 225,
        "step": 5,
        "mandatory": false
      },
      "duration": {
        "min": [
          0,
          1
        ],
        "max": [
          6,
          0
        ],
        "mandatory": false
      }
    }
  },
  {
    "programId": 31,
    "program": "Bottom heat",
    "parameters": {
      "temperature": {
        "min": 100,
        "max": 200,
        "step": 5,
        "mandatory": false
      },
      "duration": {
        "min": [
          0,
          1
        ],
        "max": [
          12,
          0
        ],
        "mandatory": false
      }
    }
  },
  {
    "programId": 25,
    "program": "Top heat",
    "parameters": {
      "temperature": {
        "min": 100,
        "max": 225,
        "step": 5,
        "mandatory": false
      },
      "duration": {
        "min": [
          0,
          1
        ],
        "max": [
          12,
          0
        ],
        "mandatory": false
      }
    }
  },
  {
    "programId": 10,
    "program": "Full grill",
    "parameters": {
      "temperature": {
        "min": 225,
        "max": 225,
        "step": null,
        "mandatory": false
      },
      "duration": {
        "min": [
          0,
          1
        ],
        "max": [
          6,
          0
        ],
        "mandatory": false
      }
    }
  },
  {
    "programId": 11,
    "program": "Economy grill",
    "parameters": {
      "temperature": {
        "min": 225,
        "max": 225,
        "step": null,
        "mandatory": false
      },
      "duration": {
        "min": [
          0,
          1
        ],
        "max": [
          6,
          0
        ],
        "mandatory": false
      }
    }
  },
  {
    "programId": 29,
    "program": "Fan grill",
    "parameters": {
      "temperature": {
        "min": 50,
        "max": 225,
        "step": 5,
        "mandatory": false
      },
      "duration": {
        "min": [
          0,
          1
        ],
        "max": [
          6,
          0
        ],
        "mandatory": false
      }
    }
  },
  {
    "programId": 18,
    "program": "Cake plus",
    "parameters": {
      "temperature": {
        "min": 30,
        "max": 225,
        "step": 5,
        "mandatory": false
      },
      "duration": {
        "min": [
          0,
          1
        ],
        "max": [
          12,
          0
        ],
        "mandatory": false
      }
    }
  },
  {
    "programId": 6,
    "program": "Eco Fan heat",
    "parameters": {
      "temperature": {
        "min": 30,
        "max": 225,
        "step": 5,
        "mandatory": false
      },
      "duration": {
        "min": [
          0,
          1
        ],
        "max": [
          12,
          0
        ],
        "mandatory": false
      }
    }
  },
  {
    "programId": 75,
    "program": "Eco Steam cooking",
    "parameters": {
      "temperature": {
        "min": 40,
        "max": 100,
        "step": 5,
        "mandatory": false
      },
      "duration": {
        "min": [
          0,
          1
        ],
        "max": [
          10,
          0
        ],
        "mandatory": false
      }
    }
  }
]

And for the Oven type H7860-60BP this is the dump: 
[
  {
    "programId": 13,
    "program": "Fan plus",
    "parameters": {
      "temperature": {
        "min": 30,
        "max": 250,
        "step": 5,
        "mandatory": false
      },
      "duration": {
        "min": [
          0,
          1
        ],
        "max": [
          12,
          0
        ],
        "mandatory": false
      }
    }
  },
  {
    "programId": 24,
    "program": "Conventional heat",
    "parameters": {
      "temperature": {
        "min": 30,
        "max": 280,
        "step": 5,
        "mandatory": false
      },
      "duration": {
        "min": [
          0,
          1
        ],
        "max": [
          12,
          0
        ],
        "mandatory": false
      }
    }
  },
  {
    "programId": 14,
    "program": "Intensive bake",
    "parameters": {
      "temperature": {
        "min": 50,
        "max": 250,
        "step": 5,
        "mandatory": false
      },
      "duration": {
        "min": [
          0,
          1
        ],
        "max": [
          12,
          0
        ],
        "mandatory": false
      }
    }
  },
  {
    "programId": 7,
    "program": "Auto roast",
    "parameters": {
      "temperature": {
        "min": 100,
        "max": 230,
        "step": 5,
        "mandatory": false
      },
      "duration": {
        "min": [
          0,
          1
        ],
        "max": [
          12,
          0
        ],
        "mandatory": false
      }
    }
  },
  {
    "programId": 31,
    "program": "Bottom heat",
    "parameters": {
      "temperature": {
        "min": 100,
        "max": 280,
        "step": 5,
        "mandatory": false
      },
      "duration": {
        "min": [
          0,
          1
        ],
        "max": [
          12,
          0
        ],
        "mandatory": false
      }
    }
  },
  {
    "programId": 25,
    "program": "Top heat",
    "parameters": {
      "temperature": {
        "min": 100,
        "max": 250,
        "step": 5,
        "mandatory": false
      },
      "duration": {
        "min": [
          0,
          1
        ],
        "max": [
          12,
          0
        ],
        "mandatory": false
      }
    }
  },
  {
    "programId": 10,
    "program": "Full grill",
    "parameters": {
      "temperature": {
        "min": 200,
        "max": 300,
        "step": 5,
        "mandatory": false
      },
      "duration": {
        "min": [
          0,
          1
        ],
        "max": [
          6,
          0
        ],
        "mandatory": false
      }
    }
  },
  {
    "programId": 11,
    "program": "Economy grill",
    "parameters": {
      "temperature": {
        "min": 200,
        "max": 300,
        "step": 5,
        "mandatory": false
      },
      "duration": {
        "min": [
          0,
          1
        ],
        "max": [
          6,
          0
        ],
        "mandatory": false
      }
    }
  },
  {
    "programId": 29,
    "program": "Fan grill",
    "parameters": {
      "temperature": {
        "min": 100,
        "max": 260,
        "step": 5,
        "mandatory": false
      },
      "duration": {
        "min": [
          0,
          1
        ],
        "max": [
          6,
          0
        ],
        "mandatory": false
      }
    }
  },
  {
    "programId": 6,
    "program": "Eco Fan heat",
    "parameters": {
      "temperature": {
        "min": 100,
        "max": 250,
        "step": 5,
        "mandatory": false
      },
      "duration": {
        "min": [
          0,
          1
        ],
        "max": [
          12,
          0
        ],
        "mandatory": false
      }
    }
  }
]

as said both I am not able to put program through api as I always get the following error:
{
  "code": 400,
  "message": "Response from setting program was not OK or OK but RESPONSE_FROM_DEVICE_ERROR instead."
}

For dishwasher and drawer it works fine... thanks for your help!

**Comment by astrandb** [comment]
This should work for your oven.
Go to services in developer tool. Select service `Miele start program` and select your oven entity. Then switch to yaml mode.
```yaml
service: miele.set_program
target:
  device_id: b579e-your-oven-device-id-14e
data:
  programId: 24
  temperature: 150
  duration: [1, 0]
```
Similar service works with my washing machine, However I can not set temperature, spin cycle etc. Only default values for the program.

Good luck!

**Reply by Ursli71** [reply]
I was offline for a few days. 
thanks for the hint, but unfortunately it did not work. The Oven I cannot get to work. Even if directly via api in swagger-ui from Miele I do get the following  error code:
{
  "code": 400,
  "message": "Response from setting program was not OK or OK but RESPONSE_FROM_DEVICE_ERROR instead."
}
So I don't think it is HA issue. My latin has come to an end and Miele does not reacted on my request to look into my 2 ovens, which I cannot control via api . So unfortunatly I guess I will for now only be able to control my dishwasher, warming drawer and fridge.  Unless you have a good idea :-)

**Comment by Ursli71** [comment]
and a big thanks for including the warming drawer in the integration. Have upgraded to the latest version, changed my templates to the correct program types and it works nicely.
